runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -197,104 +327,139 @@ impl API for Module {
     }
 }
 
-impl Module {
-    const EVM_CONFIG: Config = Config::istanbul();
+impl<Cfg: Config> Module<Cfg> {
+    const EVM_CONFIG: EVMConfig = EVMConfig::istanbul();
 
-    fn do_evm<C, F, V>(
-        source: H160,
-        value: U256,
-        gas_price_in: U256,
-        gas_limit: u64,
-        ctx: &mut C,
-        f: F,
-    ) -> Result<V, Error>
+    fn do_evm<C, F, V>(source: H160, value: U256, ctx: &mut C, f: F) -> Result<V, Error>
     where
         F: FnOnce(
             &mut StackExecutor<'static, MemoryStackState<'_, 'static, evm_backend::Backend<'_, C>>>,
+            u64,
         ) -> (evm::ExitReason, V),
-        C: Context,
+        C: TxContext,
     {
-        // Make sure that gas_price >= min_gas_price.
         let params = Self::params(ctx.runtime_state());
-        let gas_price: primitive_types::U256 = gas_price_in.into();
+        let den = params.token_denomination;
         let min_gas_price: primitive_types::U256 = params.min_gas_price.into();
+
+        let gas_limit: u64 = core::Module::remaining_tx_gas(ctx);
+        let gas_price: primitive_types::U256 = ctx.tx_auth_info().fee.gas_price().into();
+
+        // Make sure that gas_price >= min_gas_price.
         if gas_price < min_gas_price {
             return Err(Error::GasPriceTooLow);
         }
 
         let vicinity = evm_backend::Vicinity {
-            gas_price: gas_price_in,
+            gas_price: gas_price.into(),
             origin: source,
         };
 
+        // The maximum gas fee has already been withdrawn in authenticate_tx().
+        let paid_fee = ctx.tx_auth_info().fee.amount.amount();
+        let max_gas_fee = gas_price
+            .checked_mul(primitive_types::U256::from(gas_limit))
+            .ok_or(Error::FeeOverflow)?;
+
+        if paid_fee < max_gas_fee.as_u128() {
+            return Err(Error::InsufficientBalance);
+        }
+
         let mut backend = evm_backend::Backend::<'_, C>::new(vicinity, ctx);
         let metadata = StackSubstateMetadata::new(gas_limit, &Self::EVM_CONFIG);
         let stackstate = MemoryStackState::new(metadata, &backend);
         let mut executor = StackExecutor::new(stackstate, &Self::EVM_CONFIG);
 
-        let (exit_reason, exit_value) = f(&mut executor);
+        // Withdraw the value from the account.
+        executor
+            .state_mut()
+            .withdraw(source.into(), value.into())
+            .map_err(|_| Error::ValueWithdrawalFailed)?;
+
+        // Run EVM.
+        let (exit_reason, exit_value) = f(&mut executor, gas_limit);
+
+        if !exit_reason.is_succeed() {
+            return Err(Error::EVMError(format!("{:?}", exit_reason)));
+        }
 
         let gas_used = executor.used_gas();
 
         if gas_used > gas_limit {
+            core::Module::use_tx_gas(ctx, gas_limit).map_err(Error::Core)?;
```

Comment:
Hm, will this ever be triggered given that you have an `!exit_reason.is_succeed()` check and early return above?

Reply:
Probably not, but I'd still like to keep this here just in case (the comparison is cheap).